drivers/led/led_gpio.c

```diff
@@ -36,6 +38,13 @@ struct led_dts_cfg {
 struct led_gpio_data {
 	struct device *gpio_dev;
 	const struct led_dts_cfg *cfg;
+#if defined(CONFIG_LED_SOFTWARE_BLINK)
+	struct k_timer timer;
```

Comment:
It can be a problem. Maybe it was not possible to sleep in the `gpio_pin_set` or I did not realize that it is possible. Probably in that case we have to use a timer instance and workqueue instance for software blink but if this function can sleep then we cannot use also the system workqueue because it is also not good idea to sleep there so we have to define own workqueue for software blink feature which means creating the new thread for me it would be some overkill in used resource.

I will check, how Linux handle LEDs. I would like to have a possibility to call LED API from irq context.

BTW There is no any word in `gpio_pin_set` that this function should not be used from IRQ context or that it can sleep.
Do we have any cases where I2C GPIO expander is under it ?

Reply:
Ok, we have SPI/I2C GPIO expander driver which works synchronously

Reply:
I checked the Linux. Just as I thought it used combination of software timer and workqueue to handle software blink.

Reply:
So since I know `gpio_pin_set` can be blocking we probably need to do some simillar logic here or like you said we can move this to something like generic ` led.c`

Reply:
I have been thinking more to this sleeping context issue.

Here is how it could be handled:

We could introduce some _nosleep variants for LED API functions and also a "can_sleep" flag to advertise if a LED device/driver can sleep or not. If a user calls a classic variant (e.g. led_on), then it means that the driver is allowed to sleep. The driver callback can be run directly. But if a user calls the _nosleep variant (e.g. led_on_nosleep), then the LED system has to check if the driver may sleep or not. If not then again the driver callback can be run directly. Else a work queue can be used to defer the work into a thread context. A timer handler (as for software blinking) should use the _nosleep variants.

So I think it is correct to use timers for LED software blinking. We don't want the work queue overhead if we can avoid it. The "sleeping" context issue is already here and the fix should not prevent from configuring LEDs from timers.

As a side note, AFAIK the "can_sleep" information is not exported/supported by the GPIO subsystem yet. At some point it may be needed by the led-gpio driver.

Reply:
Maybe "Asynchronous sequence manager #23075" can be used for LED software blinking. cc @nordic-krch

Reply:
that wouldn't not fit here but if gpio expander would use i2c_async (#23371) then it could be called from interrupt context.

Reply:
> As a side note, AFAIK the "can_sleep" information is not exported/supported by the GPIO subsystem yet. At some point it may be needed by the led-gpio driver.

The GPIO API is documented to return `-EWOULDBLOCK` when invoked from a context where the operation can't be completed synchronously.  This is implemented in the SX1509B driver, and I believe in subsequent IOX devices.

I don't see how `i2c_sync` would help with invoking synchronous GPIO calls from an interrupt.  Though if the `-EWOULDBLOCK` return value were detected the operation could be re-initiated with an async GPIO API, but that API would have to be designed.

Seems a lot more complex than a work queue, though.

Reply:
> I don't see how i2c_sync would help with invoking synchronous GPIO calls from an interrupt

They would not be synchronous. Using work queue adds much more jitter and latency compared to timer. From on chip gpios this can be seen as unnecessary limitation.

Reply:
If the GPIO API can return the `-EWOULDBLOCK` when invoked from a context where the operation can't be completed synchronously. I would suppose that timer handler is such context then we can call it again by triggering a workqueue from timer context and there we can call the GPIO API. For pure GPIO driving the GPIO API will be called directly in timer handler because it does not return `-EWOULDBLOCK` so I think that it will not be a limitation there. From the other side it would not be a good idea to use the system workqueue because blocking there have an impact on other subsystem for example Bluetooth. We can define here a workqueue instance which will be used only by this feature but it means creating a new thread which will significantly increase the volatile memory consumption for such a simple functionality like LED blinking.

Reply:
Bluetooth should not be affected by items in the system work queue unless those items misbehave.  I don't think this driver should define a workqueue, but agree we need better control of which work queue is to be used by drivers that need them.

Reply:
Again, I think that software LED blinking should be a fallback for all the LED devices/drivers without hardware blinking capabilities, and not only GPIO LEDs.

Moreover context should be handled from the led_set_{on,off,brightness} functions. Software LED blinking is only a user among others. To make a decision between calling the LED driver's callback directly or defer it to a workqueue, this functions need to be aware of the context (interrupt, thread) and also if the driver may sleep. I don't think that calling the driver's callback and handling a -EWOULDBLOCK error is correct. GPIO is not the only backend we have to deal with. It would be better to have a "can_sleep" flag advertised by the LED drivers. But to be efficient this needs some support from all the possible backends (including GPIO)...

Reply:
>  I don't think that calling the driver's callback and handling a -EWOULDBLOCK error is correct. GPIO is not the only backend we have to deal with. It would be better to have a "can_sleep" flag advertised by the LED drivers. 

GPIO is not the only API where an operation may not be [isr-ok](https://docs.zephyrproject.org/latest/reference/terminology.html#api-term-isr-ok).  The standard way of nicely supporting code that tries something that can't be done in the calling (interrupt) context is to return `-EWOULDBLOCK`.  This fast-exit should have minimal performance impact, essentially no different to retrieving and checking a flag.

I understand that this seems a little odd, but it is the most flexible/least impact solution we have available.  Since this LED driver can't tell whether a particular GPIO has to support can change pin state from interrupt context, so there is no way it could advertise a "can sleep" flag.